sys/ktest.c

```diff
@@ -187,3 +190,33 @@ void ktest_main(const char *test) {
       ktest_atomically_print_success();
   }
 }
+
+void ktest_usermode_exit(thread_t *td, int status) {
+  assert(ktest_test_running_flag);
+
+  sleepq_signal(td);
+  td->td_ktest_status = status;
+}
+
+void ktest_usermode_timeout(void *arg) {
+  kprintf("Usermode timed out!\n");
+  ktest_failure();
+}
+
+void ktest_wait_for_user_thread(thread_t *td, int timeout_ms) {
+  callout_t timeout_callout;
+  if (timeout_ms > 0)
+    callout_setup_relative(&timeout_callout, timeout_ms, ktest_usermode_timeout,
+                           NULL);
+
+  sleepq_wait(td, "usermode ktest");
+  int status = td->td_ktest_status;
```

Comment:
If `td_ktest_status` plays a role of exit code then just name it this way (i.e. `td_exitcode`) and make it a regular member of thread structure (and not `ktest` specific).